src/v3/digest/digest.test.ts

```diff
@@ -1,41 +1,150 @@
+      ];
+      const digest = digestDocument(document, salts, []);
+      const expectedDigest = "100ffe750d2ac887d9c4def00849dc7c08b21bba4a351883a6fc57b8c12d7474";
+      expect(digest).toEqual(expectedDigest);
+    });
+  });
+
+  describe("salt", () => {
+    test("handles shadowed keys correctly (type 1: root, dot notation)", () => {
+      const document = {
+        ...cloneDeep(sampleDoc),
+        "oaProof.value": "0xSomeMaliciousDocumentStore, this would be at oaProof.value after flatMap if uncaught"
+      };
+      expect(() => {
+        salt(document);
+      }).toThrow("Key names must not have . in them");
+    });
+    test("handles shadowed keys correctly (type 2: root, array index)", () => {
+      const document = {
+        ...cloneDeep(sampleDoc),
+        "type[1]": "MaliciousCredential, this would be at type[1] after flatMap if uncaught"
+      };
+      expect(() => {
+        salt(document);
+      }).toThrow("Key names must not have '[' or ']' in them");
+    });
+    test("handles shadowed keys correctly (type 3: nested as object, dot notation)", () => {
+      const document = {
+        ...cloneDeep(sampleDoc),
+        nested: {
+          "oaProof.value":
+            "0xSomeMaliciousDocumentStore, this would be at nested.oaProof.value after flatMap if uncaught"
         }
       };
+      expect(() => {
+        salt(document);
+      }).toThrow("Key names must not have . in them");
+    });
+    test("handles shadowed keys correctly (type 4: nested as object, array index)", () => {
+      const document = {
+        ...cloneDeep(sampleDoc),
+        nested: { "type[1]": "this would be at nested.type[1] after flatMap if uncaught" }
+      };
+      expect(() => {
+        salt(document);
+      }).toThrow("Key names must not have '[' or ']' in them");
+    });
+    test("handles shadowed keys correctly (type 5: nested as array, dot notation)", () => {
+      const document = {
+        ...cloneDeep(sampleDoc),
+        nested: [{ "shadowed.key": "this would be at nested[0].shadowed.key after flatMap if uncaught" }]
+      };
+      expect(() => {
+        salt(document);
+      }).toThrow("Key names must not have . in them");
+    });
+    test("handles shadowed keys correctly (type 6: nested as array, array index)", () => {
+      const document = {
+        ...cloneDeep(sampleDoc),
+        nested: [{ "type[1]": "this would be at nested[0].type[1] after flatMap if uncaught" }]
+      };
+      expect(() => {
+        salt(document);
+      }).toThrow("Key names must not have '[' or ']' in them");
+    });
 
-      const salts = salt(document);
-      const digest = digestDocument(document, salts, []);
-      expect(digest).toBeTruthy();
+    test("handles null values correctly", () => {
+      const document = {
+        ...cloneDeep(sampleDoc),
+        grades: null
+      };
+      expect(() => {
+        salt(document);
+      }).toThrow("Cannot convert undefined or null to object");
```

Comment:
what throw this error? null is a valid json value, so should be ok :grimacing:

Reply:
```
  if (typeof value === "object") {
    return Object.keys(value).flatMap(key => deepMap(value[key], path ? `${path}.${key}` : key));
  }
```

looks like that's because the type of `null` values are `object` so the error is caused by this part of `deepMap`. since null values are accepted, we'll update the test cases for schema too as well

Reply:
ok since `null` is an accepted value, i modified `deepMap` to accept it

also changed schema tests since error messages for null values should have changed (now expecting schema errors instead)

Reply:
fixed

- since null is accepted value, changed some tests for schema
- edited deepMap type check (to handle null properly)
- improved digest test case expected values by using toContainEqual+objectContaining